### PR TITLE
compose: show detailed errors again

### LINKED ISSUE #1942: [Regression in v0.23.0] Compose does not print detailed error
```yaml
services:
  foo:
    image: hello-world
    runtime: invalid
```

v0.22.2
```
$ nerdctl compose up
INFO[0000] Ensuring image hello-world                   
INFO[0000] Re-creating container tmp_foo_1              
FATA[0000] failed to create shim task: OCI runtime create failed: unable to retrieve OCI runtime error (open /run/containerd/io.containerd.runtime.v2.task/default/88a3ee42b9cc54ee3a275b4a21eec2e610d8fd0470353bf3f89aa987acdf50c6/log.json: no such file or directory): exec: "invalid": executable file not found in $PATH: unknown 
FATA[0000] error while creating container tmp_foo_1: exit status 1
```

v0.23.0
```yaml
$ nerdctl compose up
INFO[0000] Ensuring image hello-world                   
INFO[0000] Re-creating container tmp_foo_1              
FATA[0000] error while creating container tmp_foo_1: exit status 1
```

Still not fixed in the main branch.

Seems a regression in 3296c1430679ffede4035744395ff4234296f27c

### PR DESCRIPTION
**Pull request by AkihiroSuda**
The detailed error was not printed since 3296c14 (v0.23.0).

Fix #1942

### CODE REVIEW

File: pkg/composer/up_service.go [hunk 1]
```diff
@@ -165,10 +165,15 @@ func (c *Composer) upServiceContainer(ctx context.Context, service *serviceparse
...
 	}
 
+	// Containers are always detached, even for `nerdctl compose up --detach=false`.
+	// (In that case, containers are detached and then the log viewer is "attached")
+	//
+	// See pkg/composer/serviceparser/serviceparser.go
 	if !container.Detached {
-		cmd.Stdout = os.Stdout
-		cmd.Stderr = os.Stderr
+		panic("container.Detached can't be false for compose mode")
```
Comment (djdongjin) [inline pkg/composer/up_service.go:173]: Should it return an error or panic directly?

Comment (AkihiroSuda) [reply pkg/composer/up_service.go:173]: Intentionally panicking, because unlike other errors this error must not happen with any user operation.

**Review by djdongjin — APPROVED** [review]

File: pkg/composer/up_service.go [hunk 2]
```diff
@@ -165,10 +165,15 @@ func (c *Composer) upServiceContainer(ctx context.Context, service *serviceparse
 		cmd.Stdin = os.Stdin
 	}
 
+	// Containers are always detached, even for `nerdctl compose up --detach=false`.
```
Comment (fahedouch) [inline pkg/composer/up_service.go:168]: This message is confusing. containers created with `nerdctl  run` may be detached or not but compose should always read from container std streams no matter the underlying container is detached or not . In other worder `detach` option should have no effect on the compose level but should keep the standard effect on underlying `nerdcrl run` cmd

Comment (AkihiroSuda) [reply pkg/composer/up_service.go:168]: - `nerdctl compose up --detach=true`: launches `nerdctl run -d`
- `nerdctl compose up --detach=false`: launches `nerdctl run -d` (same as above), and then "attach" `nerdctl logs`

Comment (fahedouch) [reply pkg/composer/up_service.go:168]: Detach should be true to run nerdctl with `-d`  (see https://github.com/containerd/nerdctl/blob/4d9133f76f36efac6115aee2806c67a9b2ef2e8c/pkg/composer/up_service.go#L144)

Comment (fahedouch) [reply pkg/composer/up_service.go:168]: `nerdctl compose up --detach=false` : why `nerdctl run -d` then attach the log viewer . Why not `nerdctl run` and connect cmd std streams to standard std streams ?

Comment (AkihiroSuda) [reply pkg/composer/up_service.go:168]: To allow `nerdctl compose logs` after running `nerdctl compose up --detach=false`

Comment (fahedouch) [reply pkg/composer/up_service.go:168]: OK I forgot that we enforced to write to log file even without `-d` . But I am not agreed with the fact that the underlying `nerdctl run` cmd is always running with the `-d` flag.

Comment (AkihiroSuda) [reply pkg/composer/up_service.go:168]: You're right, the `container.Detached` is always true, but `nerdctl run` isn't always launched with `-d`.

I guess we should just remove the `container.Detached` struct field.

Comment (AkihiroSuda) [reply pkg/composer/up_service.go]: Updated

**Review by manugupt1 — APPROVED** [review]

File: pkg/composer/up_service.go (hunk 2 above)
Comment (fahedouch) [reply pkg/composer/up_service.go]: that means `container.Detached` have no effect. Users can no longer drop stdout by setting `container.Detached` into the yaml definition ( I believe  it is not compatible with docker too)

I think we should keep `container.Detached` but not always true and then forward  cmd std streams to os std.

that means:

- `container.detached = true` :  `nerdctl run -d ` will log nothing => os std streams will receive nothing
- `container.detached = false` :  `nerdctl run` may log data => os std streams will receive data

```
	if container.Detached && !service.Unparsed.StdinOpen && !service.Unparsed.Tty {
		container.RunArgs = append([]string{"-d"}, container.RunArgs...)
	}

	if service.Unparsed.StdinOpen != service.Unparsed.Tty {
		return "", fmt.Errorf("currently StdinOpen(-i) and Tty(-t) should be same")
	}
	if service.Unparsed.StdinOpen {
		cmd.Stdin = os.Stdin
	}
	cmd.Stdout = os.Stdout
	cmd.Stderr = os.Stderr
```

Comment (AkihiroSuda) [reply pkg/composer/up_service.go]: `Detached` was never settable via the YAML AFAICS
https://github.com/compose-spec/compose-spec/blob/master/spec.md

Comment (fahedouch) [reply pkg/composer/up_service.go]: ok but strange! LGTM in this case

**Pull request merged**